### PR TITLE
load/store lastFlushedSeq from/to tlogStorage's metadata

### LINKED ISSUE #273: nbdserver.tlogStorage : store last flushed sequence in  metadata server
Store last sent sequence the tlogStorage's vdisk's storageCluster's metadataServer.  This will then also be loaded when loading/starting the vdisk (if possible).

Part of #230.

### PR DESCRIPTION
**Pull request by GlenDC**
fixes #273

### CODE REVIEW

File: nbd/nbdserver/tlog.go [hunk 1]
```diff
@@ -24,11 +28,16 @@ const (
 // newTlogStorage creates a tlog storage BlockStorage,
 // wrapping around a given backend storage,
 // using the given tlog client to send its write transactions to the tlog server.
-func newTlogStorage(ctx context.Context, vdiskID, clusterID string, configSource config.Source, blockSize int64, storage storage.BlockStorage, client tlogClient) (storage.BlockStorage, error) {
+func newTlogStorage(ctx context.Context, vdiskID, clusterID string, configSource config.Source, blockSize int64, storage storage.BlockStorage, mdProvider ardb.MetadataConnProvider, client tlogClient) (storage.BlockStorage, error) {
 	if storage == nil {
 		return nil, errors.New("tlogStorage requires a non-nil BlockStorage")
 	}
 
+	metadata, err := newTlogMetadata(vdiskID, mdProvider)
```
Comment (GlenDC) [inline nbd/nbdserver/tlog.go]: Here is where we load the lastFlushedSequence from the ARDB metadata storage server (if it existed, if it didn't exist `metadata.LastFlushedSequence` will equal to `0`).

File: nbd/nbdserver/tlog.go [hunk 2]
```diff
@@ -249,12 +266,50 @@ func (tls *tlogStorage) Flush() (err error) {
...
+		defer wg.Done()
+
+		// get last flushed sequnece from tlogclient
+		// NOTE: perhaps it would be better to keep the metadata up to date at all times
+		//       but for now I didn't see an easy/efficient way to keep the last flushed sequnece
+		//       up to date at all times, so I'm simply updating it right before flushing
+		tls.metadata.LastFlushedSequence = tls.tlog.LastFlushedSequence()
+
+		// flush last flushed sequence
+		err := tls.metadata.Flush()
```
Comment (GlenDC) [inline nbd/nbdserver/tlog.go]: Here is where we store the LastFlushedSequence into the ARDB metadata storage server used for this vdisk. A value is always stored, even if it is equal to `0`.

**Comment by GlenDC** [comment]
@iwanbk this PR is ready for you to review and merge.

File: nbd/nbdserver/tlog.go [hunk 3]
```diff
@@ -249,12 +266,50 @@ func (tls *tlogStorage) Flush() (err error) {
...
+
+	go func() {
+		defer wg.Done()
+
+		// get last flushed sequnece from tlogclient
+		// NOTE: perhaps it would be better to keep the metadata up to date at all times
+		//       but for now I didn't see an easy/efficient way to keep the last flushed sequnece
+		//       up to date at all times, so I'm simply updating it right before flushing
+		tls.metadata.LastFlushedSequence = tls.tlog.LastFlushedSequence()
+
```
Comment (iwanbk) [inline nbd/nbdserver/tlog.go]: This is wrong.

What we need to store is not last sequence flushed by tlog server.

But last sequence flushed by nbd server to ardb storage.

There is no point to store the former in nbd.

File: nbd/nbdserver/tlog.go [hunk 4]
```diff
@@ -864,6 +920,72 @@ func (dh *dataHistory) Empty() bool {
...
+		key = TlogMetadataKey(vdiskID)
+
+		lastFlushedSequence, err = redis.Uint64(
+			conn.Do("HGET", key, tlogMetadataLastFlushedSequenceField))
+		if err != nil && err != redis.ErrNil {
+			return nil, err
+		}
+	}
+
```
Comment (iwanbk) [inline nbd/nbdserver/tlog.go]: what if the `provider` is nil?

Comment (GlenDC) [reply nbd/nbdserver/tlog.go]: That's only ever a thing for testing purposes, so we don't require to create a provider if we don't really care about it. For production the provider given to this function is never `nil`.

**Comment by GlenDC** [comment]
@iwanbk I addressed your concerns. Please review once again if you want.

File: docs/nbd/storage/tlog.md [hunk 5]
```diff
@@ -14,6 +14,8 @@ The TLog storage delegates the actual storage work to the [deduped][deduped]- or
 
 Rather than defining actual [block][block] storage logic, the TLog storage has another purpose instead. All write (`SET`, `DELETE`, `MERGE`) operations are send asynchrounsouly to the [TLog Server][tlogserver], using the [TLog Client][tlogclient]. By doing so the [vdisk][vdisk] can be [rolled back][rollback] to an earlier [logged (3)][log] in history. Additionally the [Transaction Log][tlog] can also be used to create a new [vdisk][vdisk] using the [logged (3)][log] transactions of a TLog-supported [vdisk][vdisk].
 
+This storage does have some metadata linked to it, which it stored in the [ardb][ardb] [Metadata][metadata] Storage Server defined in its configuration. An example of such metadata is the "last flushed sequence", as in the sequence of the last block flushed by the [TLog Server][tlogserver]. The last Flushed sequence index is stored, such that we can compare it with the version the tlogserver has for this vdisk, and sync (see: [restore][restore]) with it at startup in case that is needed.
```
Comment (iwanbk) [inline docs/nbd/storage/tlog.md]: >  An example of such metadata is the "last flushed sequence", as in the sequence of the last block flushed by the [TLog Server][tlogserver]

should be last block flushed by nbd server

File: nbd/nbdserver/tlog.go [hunk 6]
```diff
@@ -616,6 +648,7 @@ type tlogClient interface {
 	WaitNbdSlaveSync() error
 	ChangeServerAddresses([]string)
 	Recv() <-chan *tlogclient.Result
+	LastFlushedSequence() uint64
```
Comment (iwanbk) [inline nbd/nbdserver/tlog.go]: Should be deleted

File: tlog/tlogclient/client.go [hunk 7]
```diff
@@ -331,6 +331,11 @@ func (c *Client) Recv() <-chan *Result {
 	return reChan
 }
 
+// LastFlushedSequence returns the last flushed sequence of this tlog client.
```
Comment (iwanbk) [inline tlog/tlogclient/client.go]: should be deleted

**Comment by GlenDC** [comment]
OK @iwanbk, I addressed your latest code review. Please review once again if you want.

**Comment by iwanbk** [comment]
LGTM

**Pull request merged**